### PR TITLE
Encrypt nonce requests with viewing key

### PR DESCRIPTION
**Pull request by BedrockSquirrel**
### Why is this change needed?

- We don't want requests to get nonce for address to be unencrypted, it could be exploitable

### What changes were made as part of this PR:

- mimic the other RPC requests that support viewing key encryption, use encrypted params on each of the hops

### What are the key areas to look at
- do the changes seem sensible in the wallet extension and the enclave

### CODE REVIEW

File: go/common/types.go [hunk 1]
```diff
@@ -29,9 +29,11 @@ type (
 	EncryptedParamsGetBalance     []byte // The params for an RPC getBalance request, as a JSON object encrypted with the public key of the enclave.
 	EncryptedParamsCall           []byte // As above, but for an RPC call request.
 	EncryptedParamsGetTxReceipt   []byte // As above, but for an RPC getTransactionReceipt request.
+	EncryptedParamsGetTxCount     []byte // As above, but for an RPC getTransactionCount request.
 	EncryptedResponseGetBalance   []byte // The response for an RPC getBalance request, as a JSON object encrypted with the viewing key of the user.
 	EncryptedResponseCall         []byte // As above, but for an RPC call request.
 	EncryptedResponseGetTxReceipt []byte // As above, but for an RPC getTransactionReceipt request.
+	EncryptedResponseGetTxCount   []byte // As above, but for an RPC getTransactionCount request.
```
Comment (joeldudley) [inline go/common/types.go:36]: @BedrockSquirrel Duplication here.

File: go/enclave/enclave.go [hunk 2]
```diff
@@ -251,14 +252,26 @@ func (e *enclaveImpl) ExecuteOffChainTransaction(encryptedParams common.Encrypte
...
-	if hs == nil {
-		return 0
+	if hs != nil {
+		// todo: we should return an error when head state is not available, but for current test situations with race
+		// 		conditions we allow it to return zero while head state is uninitialized
+		s := e.storage.CreateStateDB(hs.HeadRollup)
+		nonce = s.GetNonce(address)
+	}
+
+	encCount, err := e.rpcEncryptionManager.EncryptWithViewingKey(address, []byte(hexutil.EncodeUint64(nonce)))
```
Comment (BedrockSquirrel) [inline go/enclave/enclave.go:270]: @joeldudley turned out this was the hexutil magic function I needed, found it after you mentioned prefixing on the "0x" to the string

File: go/host/client_api_eth.go [hunk 3]
```diff
@@ -72,3 +73,18 @@ func (api *EthereumAPI) GetTransactionReceipt(_ context.Context, encryptedParams
...
+
+// GetTransactionCount returns the nonce of the wallet with the given address.
+func (api *EthereumAPI) GetTransactionCount(_ context.Context, encryptedParams common.EncryptedParamsGetTxCount) (string, error) {
+	encryptedResponse, err := api.host.EnclaveClient.GetTransactionCount(encryptedParams)
+	if err != nil {
+		return "", err
+	}
+	// todo: get rid of this hack when we stop supporting unencrypted params for testing purposes
+	// 		(the unencrypted hex number doesn't like to be bytes to hex encoded)
+	if strings.HasPrefix(string(encryptedResponse), "0x") {
```
Comment (BedrockSquirrel) [inline go/host/client_api_eth.go:85]: This seemed to me the least painful way to keep transaction injector (unencrypted nonce requests) and wallet_extension_tests (testing with viewing key enc) happy. Would gladly take pointers on a cleaner solution though :)

**Comment by BedrockSquirrel** [comment]
Closed, clean branch was easier to replay changes onto.

**Pull request closed without merging**